### DISCUSSION TITLE
Resolution scaling

### DISCUSSION THREAD

**Discussion by Chris-Mingay** [DaveGreen-Games/ShapeEngine#34] [Q&A] [answered]
First off great effort! I've been looking for a shape based engine for a while now and it's even more of a bonus that it's in C#!

Is it possible for me to have a render resolution of lets say 640 * 360 but allow the window size to be whatever I want (or fullscreen as well) ?

I note both `GameSettings` and `WindowSettings` have various options but I don't really know what to do with them!

Thanks again
Chris

**Comment by SoloByte** [comment]
> First off great effort! I've been looking for a shape based engine for a while now and it's even more of a bonus that it's in C#!

Thank you :) I hope you enjoy it!

---

> Is it possible for me to have a render resolution of lets say 640 * 360 but allow the window size to be whatever I want (or fullscreen as well) ?

This is currently not possible. I had this feature implemented in previous alpha versions of ShapeEngine but I removed it because it was complex and hard to maintain. I achieved it with render textures from raylib.

I am planning on implementing something like this again but I am not sure when and how right now...


### What do you need that feature for?
1. Do you want to achieve a pixelated look? (There is a pixelation shader available)
2. Do you always want the same "game area" independent of the screen size?
3. Something other?

I had this feature because I wanted a pixelated look but to always achieve the same pixelation I used a scaling factor rather than a set resolution.

Point 2 is a bit trickier to do and the reason why I removed it for now. If your set resolution (640x340) is the same aspect ratio as the screen / fullscreen size then it is just a matter of scaling. If it is not the same aspect ratio then I have to find the closest dimension scale it up to that and cover the other dimension with black bars. This is all possible but I have to find a good way to add it to the engine without making it too complex. (Note that mouse scaling also becomes a problem here and a few other things)

In short, I want that feature myself but I have not found a good way of implementing it yet.


---

> I note both GameSettings and WindowSettings have various options but I don't really know what to do with them!

The window settings are used to define how windows behave. Should the window be resizable or decorated, what is the default size of the window, etc. You can leave it in the default state.

The game settings are used to define the game loop and can also be left at the default state if you are unsure about it. The development resolution is used for various scaling behind the scenes, mainly automatic camera zoom scaling so that the camera always covers the same area no matter what screen/render size you have.

**Comment by Chris-Mingay** [comment]
Hey cheers for the reply, I don't have any specific agenda just yet other than trying to copy what I was previously doing with TIC-80 (which has a set resolution of 240x136). 

Potentially I was going to look into adding pixel setting and getting (again a TIC-80 thing) to this engine as a fork.

I guess to answer your question, it would be item 2) of your options. However I can give option 1 a look!

Cheers
Chris

**Reply by SoloByte** [reply]
Okay good to know! I was just curious about your use case to have some feedback on how to implement such a feature and be able to focus on certain things :)

**Comment by Chris-Mingay** [comment]
Hey Solo, just thought I'd give you a shout and show you what I'm up to with your engine. 

![image](https://github.com/user-attachments/assets/55b79c8a-7268-4894-ae85-a29b1a6b6a9a)

If there's anything in particular I can contribute with (even if it's only to discuss solutions etc) gimme a shout.

I really love being about to draw stuff within the engine, I find it way easier than drawing actual graphics (also being able to `dotnet watch` is amazing too)

**Reply by SoloByte** [reply]
Thanks, I'll keep that in mind:) 

It's the same for me and that's why I made Shape Engine👍

_I'm still surprised that there is pretty much no engine or framework out there that does this 🙃_

**Comment by SoloByte** [comment, marked as answer]
The new version will have the following screen texture modes to control how the screen texture is scaled and placed on the screen:

- **Stretch** (screen texture will always have the same size as the screen - camera target resolution can be used to keep the visible area always the same)
- **Pixelation** (the size of the screen texture will be scaled down by a pixelaltion factor - the screen texture size will always be smaller than the screen size - factor > 0 & < 1)
- **Fixed** (the screen texture size will always be the same size - the texture will always be centered on the screen with black bars if necessary)
- **Nearest Fixed** (the screen texture will always have the same aspect ratio as the screen size but stays as close as possible to the fixed dimensions)
- **Anchor / UI** (anchor position and anchor stretch values can be used to scale the screen texture and place it on the screen - useful for splitscreen or minimaps for example)
- **Custom** (scaling and placement of the screen texture is up to the user)
